### PR TITLE
Add command "exec_progbuf"

### PR DESCRIPTION
**Pull request by JanMatCodasip**
Command "exec_progbuf" allows to execute a user-specified sequence of instructions using the program buffer.

Change-Id: If3b9614129d0b6fcbc33fade29d3d60b35e52f98
Signed-off-by: Jan Matyas <jan.matyas@codasip.com>

### CODE REVIEW

**Comment by JanMatCodasip** [comment]
This `exec_progbuf` command is intended to allow users to execute their own instructions using program buffer - without the need to craft these opreations using low-level commands `riscv dmi_read` and `riscv dmi_write`.

The idea to have such utility command had already been mentioned some time ago: https://github.com/riscv/riscv-openocd/pull/678#pullrequestreview-882005735

File: doc/openocd.texi [hunk 1]
```diff
@@ -10753,6 +10753,19 @@ For details on this trigger type, see the RISC-V Debug Specification.
...
+@subsection RISC-V Program Buffer Commands
+
+Program Buffer is an optional feature of RISC-V targets - it is a mechanism that debuggers
+can use to execute sequences of arbitrary instructions (small programs) on the target.
+For details on the Program Buffer, please refer to the RISC-V Debug Specification.
+
+@deffn {Command} {riscv exec_progbuf inst1 [inst2 [... inst_n]]}
+Execute the given sequence of instructions on the target using the Program Buffer.
+The instructions @var{inst1} .. @var{inst_n} shall be specified in their binary form
+(as 32-bit intergers). The terminating @var{ebreak} instruction needs not be specified -
```
Comment (timsifive) [inline doc/openocd.texi]:
```suggestion
(as 32-bit integers). The terminating @var{ebreak} instruction needs not be specified -
```

File: doc/openocd.texi [hunk 2]
```diff
@@ -10753,6 +10753,19 @@ For details on this trigger type, see the RISC-V Debug Specification.
 Clear the type 4 trigger that was set using @command{riscv itrigger set}.
 @end deffn
 
+@subsection RISC-V Program Buffer Commands
+
+Program Buffer is an optional feature of RISC-V targets - it is a mechanism that debuggers
+can use to execute sequences of arbitrary instructions (small programs) on the target.
+For details on the Program Buffer, please refer to the RISC-V Debug Specification.
+
+@deffn {Command} {riscv exec_progbuf inst1 [inst2 [... inst_n]]}
```
Comment (timsifive) [inline doc/openocd.texi]: Is it worth describing how compressed instructions fit into the 32-bit words?

**Review by timsifive — DISMISSED** [review]
Looks good.

For my personal curiosity, what are you using this command for?

File: doc/openocd.texi (hunk 1 above)
Comment (JanMatCodasip) [reply doc/openocd.texi]: Thank you - fixed.

File: doc/openocd.texi (hunk 2 above)
Comment (JanMatCodasip) [reply doc/openocd.texi]: I believe that would be useful - I have added that information along with examples. I have made minor reword of the documentation, too.

**Comment by JanMatCodasip** [comment]
Thank you for the review.

> what are you using this command for?

We use that command mainly for testing and for execution of fences in some cases. But it is a generic command - I believe  people will find various other uses for it.

**Review by timsifive — APPROVED** [review]
Looks good!
I agree that this seems generally useful. I was just curious what specific use you have. Testing and fences makes sense to me.

**Pull request merged**